server/src/copilot/utils/lock.ts

```diff
@@ -0,0 +1,31 @@
+export class Lock {
+    private isLocked: boolean;
+    private waiting: Array<() => void>;
+
+    constructor() {
+        this.isLocked = false;
+        this.waiting = [];
```

Comment:
This looks like it is unbounded and would leak memory...?

Reply:
Fixed, introduced maxWaiting to make queue bounded to prevent memory leaks.